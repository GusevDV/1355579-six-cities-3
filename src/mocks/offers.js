export default [
  {
    title: 'Apartment 1',
    price: 100,
    pricePeriod: 'night',
    thumnnailUrl: '/img/apartment-01.jpg',
    objectType: 'apartment',
    badgeText: null
  },
  {
    title: 'Apartment 2',
    price: 200,
    pricePeriod: 'night',
    thumnnailUrl: '/img/apartment-02.jpg',
    objectType: 'apartment',
    badgeText: 'VIP'
  },
  {
    title: 'Apartment 3',
    price: 250,
    pricePeriod: 'night',
    thumnnailUrl: '/img/apartment-03.jpg',
    objectType: 'apartment',
    badgeText: 'Premium'
  },
  {
    title: 'Apartment 4',
    price: 199,
    pricePeriod: 'night',
    thumnnailUrl: '/img/apartment-04.jpg',
    objectType: 'apartment',
    badgeText: null,
  }
]
