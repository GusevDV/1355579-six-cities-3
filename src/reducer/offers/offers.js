import {transformOffers, transformOffer} from '../../helpers/api-adapters.js';
import {ActionCreator as CityActions} from "../city/city.js";

const initialState = {
  data: [],
  isLoading: false,
  isError: false,
  sortType: 0,
  hoverOffer: null,
};

const ActionType = {
  FETCH_OFFERS_START: `FETCH_OFFERS_START`,
  FETCH_OFFERS_SUCCESS: `FETCH_OFFERS_SUCCESS`,
  FETCH_OFFERS_FAILURE: `FETCH_OFFERS_FAILURE`,
  CHANGE_SORT_TYPE: `CHANGE_SORT_TYPE`,
  CHANGE_HOVER_OFFER: `CHANGE_HOVER_OFFER`,
  UPDATE_OFFER: `UPDATE_OFFER`,
};

const ActionCreator = {
  fetchOffersStart: () => ({
    type: ActionType.FETCH_OFFERS_START,
  }),
  fetchOffersSuccess: (content) => ({
    type: ActionType.FETCH_OFFERS_SUCCESS,
    payload: content,
  }),
  fetchOffersFailure: () => ({
    type: ActionType.FETCH_OFFERS_FAILURE,
  }),
  changeSortType: (type) => ({
    type: ActionType.CHANGE_SORT_TYPE,
    payload: type,
  }),
  changeHoverOffer: (offer) => ({
    type: ActionType.CHANGE_HOVER_OFFER,
    payload: offer,
  }),
  updateOffer: (offer) => ({
    type: ActionType.UPDATE_OFFER,
    payload: offer,
  })
};

const ApiCall = {
  fetchOffers: () => (dispatch, getState, api) => {
    dispatch(ActionCreator.fetchOffersStart());
    return api.get(`/hotels`)
      .then((response) => {
        dispatch(ActionCreator.fetchOffersSuccess(transformOffers(response.data)));
        dispatch(CityActions.changeCity({
          name: response.data[0].city.name,
          coords: [response.data[0].city.location.latitude, response.data[0].city.location.longitude],
          zoom: response.data[0].city.location.zoom,
        }));
      })
      .catch(() => {
        dispatch(ActionCreator.fetchOffersFailure());
      });
  },
  changeOfferFavoriteStatus: (offerId, status) => (dispatch, getState, api) => {
    const statusNumber = Number(status);
    return api.post(`/favorite/${offerId}/${statusNumber}`)
      .then((response) => {
        dispatch(ActionCreator.updateOffer(transformOffer(response.data)));
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.FETCH_OFFERS_START:
      return Object.assign({}, state, {isLoading: true, isError: false});
    case ActionType.FETCH_OFFERS_SUCCESS:
      return Object.assign({}, state, {data: action.payload.slice(), isError: false, isLoading: false});
    case ActionType.FETCH_OFFERS_FAILURE:
      return Object.assign({}, state, {data: [], isError: true, isLoading: false});
    case ActionType.UPDATE_OFFER:
      return Object.assign({}, state, {
        data: state.data.map((offer) => offer.id === action.payload.id ? action.payload : offer)
      });
    case ActionType.CHANGE_SORT_TYPE:
      return Object.assign({}, state, {sortType: action.payload});
    case ActionType.CHANGE_HOVER_OFFER:
      return Object.assign({}, state, {hoverOffer: action.payload});
  }

  return state;
};

export {ActionCreator, ActionType, ApiCall, reducer};
