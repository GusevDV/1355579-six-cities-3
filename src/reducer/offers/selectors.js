import {createSelector} from 'reselect';
import {SortIndex} from '../../const.js';

const getOffers = (state) => state.offers.data;
const getCurrentCity = (state) => state.city.currentCity;

export const getOffersCurrentCity = createSelector(
    getOffers,
    getCurrentCity,
    (offers, city) => (city ? offers.filter((offer) => (offer.city === city)) : offers)
);

const getSortType = (state) => state.offers.sortType;

export const getCurrentSort = createSelector(
    getSortType,
    (sort) => sort
);

const getSortedOffers = (offers, sortType) => {
  switch (sortType) {
    case SortIndex.PRICE_LOW_TO_HIGH:
      return offers.slice().sort((a, b) => b.price - a.price);
    case SortIndex.PRICE_HIGH_TO_LOW:
      return offers.slice().sort((a, b) => a.price - b.price);
    case SortIndex.TOP_RATED_FIRST:
      return offers.slice().sort((a, b) => b.rating - a.rating);
    default:
      return offers;
  }
};
export const getOffersWithSort = createSelector(
    getOffersCurrentCity,
    getCurrentSort,
    (offers, sortType) => getSortedOffers(offers, sortType)
);

const getOfferId = (state, props) => props.offerId;

export const getCurrentOffer = createSelector(
    getOffers,
    getOfferId,
    (offers, offerId) => offers.find((offer) => offer.id === Number.parseInt(offerId, 10))
);

export const getHoverOffer = createSelector(
    getOffersWithSort,
    (state) => state.offers.hoverOffer,
    (offers, hoverOffer) => hoverOffer ? offers.find((offer) => offer.id === hoverOffer) : null
);

