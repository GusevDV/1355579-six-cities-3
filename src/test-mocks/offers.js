
const CityNames = {
  AMSTERDAM: `Amsterdam`,
  PARIS: `Paris`
};
export default [
  {
    id: 1,
    title: `Apartment 1`,
    price: 100,
    pricePeriod: `night`,
    thumnnailUrl: `/img/apartment-01.jpg`,
    objectType: `apartment`,
    isPremium: false,
    rating: 3.5,
    city: CityNames.AMSTERDAM,
    cityCoords: [
      52.37454,
      4.897976
    ],
    cityZoom: 13,
    coords: [52.3909553943508, 4.85309666406198],
  },
  {
    id: 2,
    title: `Apartment 2`,
    price: 200,
    pricePeriod: `night`,
    thumnnailUrl: `/img/apartment-02.jpg`,
    objectType: `apartment`,
    isPremium: true,
    rating: 4,
    city: CityNames.AMSTERDAM,
    cityCoords: [
      52.37454,
      4.897976
    ],
    cityZoom: 13,
    coords: [52.369553943508, 4.85309666406198],
  },
  {
    id: 3,
    title: `Apartment 3`,
    price: 250,
    pricePeriod: `night`,
    thumnnailUrl: `/img/apartment-01.jpg`,
    objectType: `apartment`,
    isPremium: true,
    rating: 2,
    city: CityNames.AMSTERDAM,
    cityCoords: [
      52.37454,
      4.897976
    ],
    cityZoom: 13,
    coords: [52.3909553943508, 4.929309666406198]
  },
  {
    id: 4,
    title: `Apartment 4`,
    price: 199,
    pricePeriod: `night`,
    thumnnailUrl: `/img/apartment-02.jpg`,
    objectType: `apartment`,
    isPremium: false,
    rating: 3,
    city: CityNames.AMSTERDAM,
    cityCoords: [
      52.37454,
      4.897976
    ],
    cityZoom: 13,
    coords: [52.3809553943508, 4.939309666406198]
  },
  {
    id: 5,
    title: `Apartment 5`,
    price: 199,
    pricePeriod: `night`,
    thumnnailUrl: `/img/apartment-02.jpg`,
    objectType: `apartment`,
    isPremium: false,
    rating: 5,
    city: CityNames.PARIS,
    cityCoords: [
      48.85661,
      2.351499
    ],
    cityZoom: 13,
    coords: [48.8647553943508, 2.349009666406198]
  }
];
