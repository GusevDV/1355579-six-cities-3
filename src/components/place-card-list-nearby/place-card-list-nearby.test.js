import React from "react";
import renderer from "react-test-renderer";
import offers from '../../test-mocks/offers.js';
import PlaceCardListNearby from "./place-card-list-nearby.jsx";
import {BrowserRouter} from 'react-router-dom';

it(`Should PlaceCardListNearby component render correctly`, () => {

  const render = renderer.create(
      <BrowserRouter>
        <PlaceCardListNearby
          nearbyOffers={offers.slice(0, 3)}
          onFavoriteClick={()=>{}}
        />
      </BrowserRouter>
  )
  .toJSON();

  expect(render).toMatchSnapshot();
});
