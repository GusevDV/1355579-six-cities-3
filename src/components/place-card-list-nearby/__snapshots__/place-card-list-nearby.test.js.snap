// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should PlaceCardListNearby component render correctly 1`] = `
<div
  className="container"
>
  <section
    className="near-places places"
  >
    <h2
      className="near-places__title"
    >
      Other places in the neighbourhood
    </h2>
    <div
      className="near-places__list places__list"
    >
      <a
        className="near-places__card place-card"
        href="/offer/1"
        onClick={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="near-places__image-wrapper place-card__image-wrapper"
        >
          <img
            alt="Place image"
            className="place-card__image"
            height="200"
            src="/img/apartment-01.jpg"
            width="260"
          />
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                100
              </b>
              <span
                className="place-card__price-text"
              >
                / 
                night
              </span>
            </div>
            <button
              className="button place-card__bookmark-button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height="19"
                width="18"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "80%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <span>
              Apartment 1
            </span>
          </h2>
          <p
            className="place-card__type"
          >
            apartment
          </p>
        </div>
      </a>
      <a
        className="near-places__card place-card"
        href="/offer/2"
        onClick={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="place-card__mark"
        >
          <span>
            Premium
          </span>
        </div>
        <div
          className="near-places__image-wrapper place-card__image-wrapper"
        >
          <img
            alt="Place image"
            className="place-card__image"
            height="200"
            src="/img/apartment-02.jpg"
            width="260"
          />
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                200
              </b>
              <span
                className="place-card__price-text"
              >
                / 
                night
              </span>
            </div>
            <button
              className="button place-card__bookmark-button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height="19"
                width="18"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "80%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <span>
              Apartment 2
            </span>
          </h2>
          <p
            className="place-card__type"
          >
            apartment
          </p>
        </div>
      </a>
      <a
        className="near-places__card place-card"
        href="/offer/3"
        onClick={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="place-card__mark"
        >
          <span>
            Premium
          </span>
        </div>
        <div
          className="near-places__image-wrapper place-card__image-wrapper"
        >
          <img
            alt="Place image"
            className="place-card__image"
            height="200"
            src="/img/apartment-01.jpg"
            width="260"
          />
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                250
              </b>
              <span
                className="place-card__price-text"
              >
                / 
                night
              </span>
            </div>
            <button
              className="button place-card__bookmark-button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height="19"
                width="18"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "40%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <span>
              Apartment 3
            </span>
          </h2>
          <p
            className="place-card__type"
          >
            apartment
          </p>
        </div>
      </a>
    </div>
  </section>
</div>
`;
